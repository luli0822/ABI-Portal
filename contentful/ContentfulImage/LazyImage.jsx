import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';

// Use this component to load images when you are not using Lazyload.

function LazyImage(props) {
	const { file, baseSizes, className, alt, onLoad, imgContainerRef } = props;

	const imgRef = useRef();
	const containerWidth = imgContainerRef.current.offsetWidth;
	useEffect(() => {
		const img = imgRef.current;
		if (img && img.complete) {
			onLoad();
		}
	});

	// on load, find the width of the container and create the picture with one img tag and one source tag for webp. Create the URL for the nearest largest 100px
	const imgWidth = baseSizes.find(size => size >= containerWidth) || 1440;
	return (
		<picture>
			<source type="image/webp" srcSet={`${file.url}?fm=webp&w=${imgWidth}`} />
			<img
				src={`${file.url}?w=${imgWidth}`}
				className={className}
				ref={imgRef}
				alt={alt}
				onLoad={onLoad}
			/>
		</picture>
	);
}

LazyImage.propTypes = {
	file: PropTypes.shape({
		url: PropTypes.string.isRequired,
	}).isRequired,
	baseSizes: PropTypes.arrayOf(PropTypes.number).isRequired,
	className: PropTypes.string.isRequired,
	alt: PropTypes.string.isRequired,
	onLoad: PropTypes.func.isRequired,
	imgContainerRef: PropTypes.shape({
		current: PropTypes.shape(),
	}).isRequired,
};

export default LazyImage;
