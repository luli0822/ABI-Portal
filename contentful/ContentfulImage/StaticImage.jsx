import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';

// Use this component to load images when you are not using Lazyload

function StaticImage(props) {
	const { file, baseSizes, className, alt, onLoad, isSVG, previewAsset } = props;

	const imgRef = useRef();

	const svgImgUrl = previewAsset || file.url;

	const imgUrl = isSVG ? svgImgUrl : `${file.url}?w=1440`;

	useEffect(() => {
		const img = imgRef.current;
		if (img && img.complete) {
			onLoad();
		}
	});
	// render out source tags for widths between 100 and 1440 px, rounded up to nearest 100px.
	return (
		<picture>
			{!isSVG && (
				<>
					{baseSizes.map((size, index) => (
						// eslint-disable-next-line react/no-array-index-key
						<React.Fragment key={index}>
							<source
								type="image/webp"
								srcSet={`${file.url}?fm=webp&w=${size}`}
								media={`(max-width: ${size}px)`}
							/>
							<source
								type={file.contentType}
								srcSet={`${file.url}?w=${size}`}
								media={`(max-width: ${size}px)`}
							/>
						</React.Fragment>
					))}
					<source
						type="image/webp"
						srcSet={`${file.url}?fm=webp&w=1440`}
						media="(min-width: 1440px)"
					/>
					<source
						type={file.contentType}
						srcSet={`${file.url}?w=1440`}
						media="(min-width: 1440px)"
					/>
				</>
			)}
			{/* img tag holds largest image src */}
			<img src={imgUrl} className={className} ref={imgRef} alt={alt} onLoad={onLoad} />
		</picture>
	);
}

StaticImage.propTypes = {
	file: PropTypes.shape({
		url: PropTypes.string.isRequired,
		contentType: PropTypes.string.isRequired,
	}).isRequired,
	baseSizes: PropTypes.arrayOf(PropTypes.number).isRequired,
	className: PropTypes.string.isRequired,
	alt: PropTypes.string.isRequired,
	onLoad: PropTypes.func.isRequired,
	isSVG: PropTypes.bool,
	previewAsset: PropTypes.string,
};

StaticImage.defaultProps = {
	isSVG: false,
	previewAsset: undefined,
};

export default StaticImage;
