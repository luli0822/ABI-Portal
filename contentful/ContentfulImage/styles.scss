.contentful-image {
  position: relative;
  height: 100%;

  &__img {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    width: 100%;
    opacity: 0;
    transition: opacity 0.5s;

    &--visible {
      opacity: 1;
    }

    &--cover {
      height: 100%;
      object-fit: cover;
    }
  }

  &__placeholder {
    filter: blur(5px);
    background-size: cover;
    background-position: center;
    transition: opacity 0.5s;
    transition-delay: 0.25s;
    height: 100%;

    &--hidden {
      opacity: 0;
    }

    &--cover {
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
    }
  }
}
