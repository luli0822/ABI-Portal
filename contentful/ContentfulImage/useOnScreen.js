import { useState, useEffect } from 'react';

// hook that uses intersection observer
export default function useOnScreen(ref, rootMargin = '0px') {
	const [isIntersecting, setIsIntersecting] = useState(false);

	useEffect(() => {
		// fallback to just load everything if IntersectionObserver is not available
		if (
			!('IntersectionObserver' in window) ||
			!('IntersectionObserverEntry' in window) ||
			!('isIntersecting' in window.IntersectionObserverEntry.prototype)
		) {
			setIsIntersecting(true);
			return () => {};
		}

		const element = ref.current;
		const observer = new IntersectionObserver(
			([entry]) => {
				setIsIntersecting(entry.isIntersecting);
			},
			{ rootMargin }
		);
		if (ref.current) {
			observer.observe(element);
		}
		return () => {
			observer.unobserve(element);
		};
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, []); // empty array because we only want to run on mount/unmount

	return isIntersecting;
}
