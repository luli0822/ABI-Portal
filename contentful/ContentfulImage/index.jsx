import React, { useContext, useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { assetPropType } from '../prop-types';
import siteDataContext from '../siteDataContext';
import StaticImage from './StaticImage';
import LazyImage from './LazyImage';
import useOnScreen from './useOnScreen';
import './styles.scss';

const rootClass = 'contentful-image';

export default function ContentfulImage({ asset, className, alt, lazyload, cover }) {
	const { lang, fieldOrFallback, previewAssets } = useContext(siteDataContext);
	const file = fieldOrFallback(asset.fields.file, lang);
	const isSVG = file && file.contentType === 'image/svg+xml';
	const [isVisible, setIsVisible] = useState(false);
	const imgContainerRef = useRef();

	const isInViewport = useOnScreen(imgContainerRef, '100px');

	const isLazyLoad = isSVG ? false : lazyload;

	function onLoad() {
		setIsVisible(true);
	}

	// Generate and array from 100px-1440px
	const getBaseSizes = () => {
		const base = 100;
		const sizes = [];
		for (let i = base; i <= 1440; i += base) {
			if (i === 1400) {
				sizes.push(1440);
			} else {
				sizes.push(i);
			}
		}
		return sizes;
	};

	//  An array of the px sizes 100px-1440
	const baseSizes = getBaseSizes();

	// Alt must be present, so we can't let it be `undefined`
	// use image description or fallback to empty string (don't pass undefined)
	const defaultAlt = fieldOrFallback(asset.fields.description) || '';
	// if no alt was passed in (default prop value is undefined, use the defaultAlt, otherwise use the passed alt, even if it's an empty string)
	const finalAlt = alt === undefined ? defaultAlt : alt;

	const placeholderClasses = [
		`${rootClass}__placeholder`,
		`${rootClass}__placeholder--${!isVisible ? 'visible' : 'hidden'}`,
		cover ? `${rootClass}__placeholder--cover` : '',
	];

	const imgClasses = [
		`${rootClass}__img`,
		`${rootClass}__img--${isVisible || isSVG ? 'visible' : 'hidden'}`,
		cover ? `${rootClass}__img--cover` : '',
	];

	const placeholderClass = placeholderClasses.join(' ');
	const imgClass = imgClasses.join(' ');

	const LazyLoadImageOn = isInViewport && (
		<LazyImage
			imgContainerRef={imgContainerRef}
			file={file}
			className={imgClass}
			baseSizes={baseSizes}
			alt={finalAlt}
			onLoad={onLoad}
		/>
	);

	const LazyLoadImageOff = (
		<StaticImage
			file={file}
			className={imgClass}
			baseSizes={baseSizes}
			alt={finalAlt}
			onLoad={onLoad}
			isSVG={isSVG}
			previewAsset={previewAssets[asset.sys.id].url}
		/>
	);

	const backgroundUrl =
		!isSVG && previewAssets[asset.sys.id] ? `url(${previewAssets[asset.sys.id].url})` : undefined;

	return (
		<div ref={imgContainerRef} className={`${rootClass} ${className}`}>
			<div
				className={placeholderClass}
				style={{
					paddingTop:
						file && !cover && `${100 * (file.details.image.height / file.details.image.width)}%`,
					backgroundImage: backgroundUrl,
				}}
			/>
			{isLazyLoad ? LazyLoadImageOn : LazyLoadImageOff}
		</div>
	);
}

ContentfulImage.propTypes = {
	asset: assetPropType.isRequired,
	className: PropTypes.string,
	alt: PropTypes.string,
	lazyload: PropTypes.bool,
	cover: PropTypes.bool,
};

ContentfulImage.defaultProps = {
	className: '',
	alt: undefined,
	lazyload: true,
	cover: false,
};
