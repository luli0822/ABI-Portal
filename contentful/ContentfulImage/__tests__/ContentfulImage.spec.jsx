import React from 'react';
import { mount, shallow } from 'enzyme';

// import useOnScreen from '../ContentfulImage/useOnScreen';
import {
	mockLang,
	mockLinkedAssets,
	mockPreviewAssets,
	mockFieldOrFallback,
} from '../../__mocks__/SiteDataContext';

// Mock React useContext

jest.mock('react', () => {
	const ActualReact = require.requireActual('react');
	return {
		...ActualReact,
		useContext: () => ({
			fieldOrFallback: mockFieldOrFallback,
			lang: mockLang,
			previewAssets: mockPreviewAssets,
		}),
	};
});

const mockImageAsset = mockLinkedAssets['test-img.png'];

const lazyloadProps = {
	asset: mockImageAsset,
	alt: 'mock-alt',
	lazyload: true,
	cover: true,
};

const noLazyloadProps = {
	asset: mockImageAsset,
	alt: 'mock-alt',
	lazyload: false,
	cover: false,
};

// const useOnScreen = jest.fn();

jest.mock('../useOnScreen');

// eslint-disable-next-line import/first
import useOnScreen from '../useOnScreen';

// eslint-disable-next-line import/first
import ContentfulImage from '..';

afterEach(() => {
	mockFieldOrFallback.mockClear();
	useOnScreen.mockClear();
});

describe('<ContentfulImage /> and <StaticImage />', () => {
	it('Renders a <LazyImage /> if is in viewport and lazyload is enabled', () => {
		useOnScreen.mockReturnValue(true);
		// eslint-disable-next-line react/jsx-props-no-spreading
		const component = shallow(<ContentfulImage {...lazyloadProps}>Page Link Text</ContentfulImage>);

		expect(component.debug()).toMatchSnapshot();
	});

	it('Does not render a <LazyImage /> if not in viewport and lazyload is enabled', () => {
		useOnScreen.mockReturnValue(false);
		// eslint-disable-next-line react/jsx-props-no-spreading
		const component = shallow(<ContentfulImage {...lazyloadProps}>Page Link Text</ContentfulImage>);

		expect(component.debug()).toMatchSnapshot();
	});

	it('Renders a <StaticImage> if lazyload is disabled, <StaticImage> renders appropriate <source> tags', () => {
		const component = mount(
			// eslint-disable-next-line react/jsx-props-no-spreading
			<ContentfulImage {...noLazyloadProps}>Page Link Text</ContentfulImage>
		);

		expect(component.debug()).toMatchSnapshot();
	});

	it("Uses the asset's provided description if no alt is provided", () => {
		const { alt, ...otherProps } = noLazyloadProps;
		const component = mount(
			// eslint-disable-next-line react/jsx-props-no-spreading
			<ContentfulImage {...otherProps}>Page Link Text</ContentfulImage>
		);

		expect(component.find("img[alt='asset description']")).toHaveLength(1);
	});

	it('Uses the empty string provided as an alt', () => {
		const { alt, ...otherProps } = noLazyloadProps;
		const component = mount(
			// eslint-disable-next-line react/jsx-props-no-spreading
			<ContentfulImage {...otherProps} alt="">
				Page Link Text
			</ContentfulImage>
		);

		expect(component.find("img[alt='']")).toHaveLength(1);
	});

	it('Renders an empty alt tag if no alt is provided and asset description is empty', () => {
		const component = mount(
			// eslint-disable-next-line react/jsx-props-no-spreading
			<ContentfulImage asset={mockLinkedAssets['test-img2.png']} lazyload={false} cover={false}>
				Page Link Text
			</ContentfulImage>
		);

		expect(component.find("img[alt='']")).toHaveLength(1);
	});
});
