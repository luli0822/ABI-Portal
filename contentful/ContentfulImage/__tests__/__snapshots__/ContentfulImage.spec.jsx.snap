// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentfulImage /> and <StaticImage /> Does not render a <LazyImage /> if not in viewport and lazyload is enabled 1`] = `
"<div className=\\"contentful-image \\">
  <div className=\\"contentful-image__placeholder contentful-image__placeholder--visible contentful-image__placeholder--cover\\" style={{...}} />
</div>"
`;

exports[`<ContentfulImage /> and <StaticImage /> Renders a <LazyImage /> if is in viewport and lazyload is enabled 1`] = `
"<div className=\\"contentful-image \\">
  <div className=\\"contentful-image__placeholder contentful-image__placeholder--visible contentful-image__placeholder--cover\\" style={{...}} />
  <LazyImage imgContainerRef={{...}} file={{...}} className=\\"contentful-image__img contentful-image__img--hidden contentful-image__img--cover\\" baseSizes={{...}} alt=\\"mock-alt\\" onLoad={[Function: onLoad]} />
</div>"
`;

exports[`<ContentfulImage /> and <StaticImage /> Renders a <StaticImage> if lazyload is disabled, <StaticImage> renders appropriate <source> tags 1`] = `
"<ContentfulImage asset={{...}} alt=\\"mock-alt\\" lazyload={false} cover={false} className=\\"\\">
  <div className=\\"contentful-image \\">
    <div className=\\"contentful-image__placeholder contentful-image__placeholder--visible \\" style={{...}} />
    <StaticImage file={{...}} className=\\"contentful-image__img contentful-image__img--hidden \\" baseSizes={{...}} alt=\\"mock-alt\\" onLoad={[Function: onLoad]} isSVG={false} previewAsset=\\"data:image/png;base64,someUriData\\">
      <picture>
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=100\\" media=\\"(max-width: 100px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=100\\" media=\\"(max-width: 100px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=200\\" media=\\"(max-width: 200px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=200\\" media=\\"(max-width: 200px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=300\\" media=\\"(max-width: 300px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=300\\" media=\\"(max-width: 300px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=400\\" media=\\"(max-width: 400px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=400\\" media=\\"(max-width: 400px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=500\\" media=\\"(max-width: 500px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=500\\" media=\\"(max-width: 500px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=600\\" media=\\"(max-width: 600px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=600\\" media=\\"(max-width: 600px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=700\\" media=\\"(max-width: 700px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=700\\" media=\\"(max-width: 700px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=800\\" media=\\"(max-width: 800px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=800\\" media=\\"(max-width: 800px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=900\\" media=\\"(max-width: 900px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=900\\" media=\\"(max-width: 900px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=1000\\" media=\\"(max-width: 1000px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=1000\\" media=\\"(max-width: 1000px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=1100\\" media=\\"(max-width: 1100px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=1100\\" media=\\"(max-width: 1100px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=1200\\" media=\\"(max-width: 1200px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=1200\\" media=\\"(max-width: 1200px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=1300\\" media=\\"(max-width: 1300px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=1300\\" media=\\"(max-width: 1300px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=1440\\" media=\\"(max-width: 1440px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=1440\\" media=\\"(max-width: 1440px)\\" />
        <source type=\\"image/webp\\" srcSet=\\"/test-img.png?fm=webp&w=1440\\" media=\\"(min-width: 1440px)\\" />
        <source type=\\"image/png\\" srcSet=\\"/test-img.png?w=1440\\" media=\\"(min-width: 1440px)\\" />
        <img src=\\"/test-img.png?w=1440\\" className=\\"contentful-image__img contentful-image__img--hidden \\" alt=\\"mock-alt\\" onLoad={[Function: onLoad]} />
      </picture>
    </StaticImage>
  </div>
</ContentfulImage>"
`;
